### ISSUE TITLE
Finetuning

### ISSUE THREAD

**Issue by splunker691** [resloved/RWKV-notebooks#3]
Testing out the finetuning. Not sure whether to expect steps/training loss or if I'm stuck with an error? 

Using:
Notebook: RWKV_LoRA.ipynb
Model: RWKV-4-Pile-1B5
Small training file: 
Raw length = 762970
Tokenized length = 190568

Colab Notebook: Tesla T4 

Current output is:

initializing deepspeed distributed: GLOBAL_RANK: 0, MEMBER: 1/1
Enabling DeepSpeed BF16.
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
Using /root/.cache/torch_extensions/py39_cu118 as PyTorch extensions root...

Flashing cursor... should I be seeing training step and loss? Or is the flashing cursor what is expected?

Tried smaller model - same results.

**Comment by resloved** [comment]
You should be able to see the epochs and loss as it trains. If you `!rm -r /root/.cache` before running the training step does it work?

**Comment by splunker691** [comment]
I will give this a try and let you know if it resolves the issue.

DIFFERENT NOTEBOOK
I tried a different notebook in the
meantime RWKV_v4_RNN_Pile_Fine_Tuning.ipynb
It was set for 100 epochs, but went past that number and I had to stop it
manually. Would it have stopped on its own or was that an error?
It did show the epochs and loss during training, but it didn't save the
vocab.json or the merges.txt

CHECKPOINTS
It did save the checkpoint .pth files
Is there any way to restart from a checkpoint and then save the finetuned
model with the .json and .txt files that are needed?

Thank you.

**Comment by resloved** [comment]
I'll try to answer your questions, but they go a bit out of the scope of why I made these notebooks.

1. I'd recommend the `v4neo` notebook over `v4` as it should be more stable.
2. Not sure what's up with it continuing after the set number of epochs. Ending it early doesn't hurt as long as it's already saved the checkpoint.
3. I believe the `vocab.json` and `merges.txt` you are referring to are from training your own models from scratch with BPE? The notebooks are assuming that you'll be fine tuning one of the models trained by BlinkDL using the Pile dataset that use `20B_tokenizer.json`.
4. If you'd like to start from a checkpoint you could always just edit the final training step to use the checkpoint path as apposed to the `$base_model`. Probably consider changing the `epoch_begin` flag to wherever you left off at.

**Comment by splunker691** [comment]
Okay, thank you, I will give these a try!
